Updating release instructions (Lombiq Technologies: OCORE-223) (#17499)
OrchardCMS/OrchardCore: .github/ISSUE_TEMPLATE/patch_release.md

## Step 1: Backporting Pull Requests

1. **Identify Pull Requests**: Review any pull requests (PRs) that need to be backported to the release branch. These should be PRs that contain strictly bug fixes and nothing else.
2. **Backport Pull Requests**: For PRs merged into the `main` branch that need to be applied to the release branch (e.g., `release/2.1.x`), comment on the merged PR with `/backport to release/2.1.x`. This comment will trigger a GitHub Action to create a new PR with the same changes for the `release/2.1.x` branch.
3. **Merge PRs**: Once all necessary PRs are created, merge them into the `release/2.1.x` branch.

## Step 2: Code and Documentation Updates

> - [ ] From the release branch (e.g., `release/2.1.x`), create a new temporary branch for the version (e.g., `release/2.1.1`).

Author: The second branch name is for the patch release. It's deleted after completing it. Previously, the guide contained `release-notes/2.1.1` for the same.

- [ ] Update version references in the documentation and code. Refer to [this PR](https://github.com/OrchardCMS/OrchardCore/pull/17065/files) for an example. The easiest is to start with a search & replace on the previous version. Version Updates Checklist:

>   - **Update `OrchardCore.Commons.props`**: Set `<VersionPrefix></VersionPrefix>` to the new version you're preparing for release.

Author: You don't actually need to set the `VersionSuffix`, ever (I don't really know why, though). It wasn't set for the last few releases and the version number correctly doesn't include it.

Reviewer: You don't, but you should. If you follow the step, we should release a preview of this version to test it out, when you do this the preview CLI will create a preview on cloud smith with the correct version instead of creating a preview with the old version.

Author: That's not what I've seen. While `VersionSuffix` was kept as `preview` throughout, the preview package published to Cloudsmith (run for the version branch), used the version number `2.1.6-preview-18513`, see [here](https://github.com/OrchardCMS/OrchardCore/actions/runs/13425099869/job/37506426047). And when published to NuGet, the workflow generated `2.1.6` see [here](https://github.com/OrchardCMS/OrchardCore/actions/runs/13425524908/job/37507630862). So, the existence of `VersionSuffix` doesn't affect the non-preview release.

Note that I'm talking about Version**Suffix**. What you have to set, and what the guide now includes, is Version**Prefix**, what I've changed to `2.1.6`.

Reviewer: okay.

## Step 1: Backporting Pull Requests

> 1. **Identify Pull Requests**: Review any pull requests (PRs) that need to be backported to the release branch. These should be PRs that contain strictly bug fixes and nothing else.

Author: This would mean that the bug fixes are strict. But the original is neither correct, should be "strictly contain...".

2. **Backport Pull Requests**: For PRs merged into the `main` branch that need to be applied to the release branch (e.g., `release/2.1`), comment on the merged PR with `/backport to release/2.1`. This comment will trigger a GitHub Action to create a new PR with the same changes for the `release/2.1` branch.
3. **Merge PRs**: Once all necessary PRs are created, merge them into the `release/2.1` branch.

## Step 2: Code and Documentation Updates

> - [ ] From the release branch (e.g., `release/2.1`), create a new temporary branch for the version (e.g., `release/2.1.1`).

Reviewer: ```suggestion
- [ ] From the release branch (e.g., `release/2.1`), create a new temporary branch for the version (e.g., `release-notes/2.1.1`).
```
This was purposely set to release-notes to distinguish the docs branches from an actual release branch.

Author: But it's not just about the release notes. There are other preparations as well, that the guide included already (changing the version numbers in the code, updating the guides), that need to happen in some branch and shouldn't happen directly in the release one.

Reviewer: yes but these are all administrative changes. I am fine with `release-notes`, if you want to change it, please change it to something other than `release` this way we don't confuse it with an actual release branch.

Author: `release-prep`?

Reviewer: Sure

Author: That's in the template now.

> - [ ] Update version references in the documentation and code. Refer to [this PR](https://github.com/OrchardCMS/OrchardCore/pull/17065/files) for an example. The easiest is to start with a search & replace on the previous version. Version Updates Checklist:

Author: Something needs to be between "search & replace" and "the previous version". I refined it a bit.

  - **Update `OrchardCore.Commons.props`**: Set `<VersionPrefix></VersionPrefix>` to the new version you're preparing for release.
  - **Update Module Versions**: Modify `src/OrchardCore/OrchardCore.Abstractions/Modules/Manifest/ManifestConstants.cs` to reflect the new version.
  - **Release Notes**: Finalize the release notes in the documentation, including:
      - Highlights and goals of the release.
      - Prerequisites for running the new version.
      - Upgrade steps and any breaking changes.
  - **Update Documentation Navigation**: Add the release notes page to `mkdocs.yml` navigation and remove it from `not_in_nav`.
  - **Version Mentions**: Update all references to the new version throughout the documentation, including:
    - [Status in the root README](https://docs.orchardcore.net/en/latest/#status)
    - CLI templates and commands.
    - Relevant guides, such as the [Creating a new decoupled CMS Website](https://docs.orchardcore.net/en/latest/guides/decoupled-cms/) guide.

> - [ ] Create a **version PR** titled `Release <version number` (e.g., `Release 2.1.1`) from the version branch (e.g., `release/2.1.1`) into the release branch (e.g., `release/2.1`)

Reviewer: ```suggestion
- [ ] Create a **version PR** titled `Release <version number>` (e.g., `Release 2.1.1`) from the version branch (e.g., `release/2.1.1`) into the release branch (e.g., `release/2.1`)
```

>   - **Update Module Versions**: Modify `src/OrchardCore/OrchardCore.Abstractions/Modules/Manifest/ManifestConstants.cs` to reflect the new version.
>   - **Release Notes**: Finalize the release notes in the documentation, including:
>       - Highlights and goals of the release.
>       - Prerequisites for running the new version.
>       - Upgrade steps and any breaking changes.
>   - **Update Documentation Navigation**: Add the release notes page to `mkdocs.yml` navigation and remove it from `not_in_nav`.
>   - **Version Mentions**: Update all references to the new version throughout the documentation, including:
>     - [Status in the root README](https://docs.orchardcore.net/en/latest/#status)
>     - CLI templates and commands.
>     - Relevant guides, such as the [Creating a new decoupled CMS Website](https://docs.orchardcore.net/en/latest/guides/decoupled-cms/) guide.
> - [ ] Create a **version PR** titled `Release <version number>` (e.g., `Release 2.1.1`) from the version branch (e.g., `release/2.1.1`) into the release branch (e.g., `release/2.1`)
> - [ ] In GitHub, manually run the `Preview - CI` workflow on your branch (NOT `main`). This will release a new preview version on Cloudsmith for testing.

Reviewer 2: ```suggestion
  - **Update Module Versions**: Modify `src/OrchardCore/OrchardCore.Abstractions/Modules/Manifest/ManifestConstants.cs` to reflect the new version.
  - **Release Notes**: Finalize the release notes in the documentation, including:
      - Highlights and goals of the release.
      - Prerequisites for running the new version.
      - Upgrade steps and any breaking changes.
  - **Update Documentation Navigation**: Add the release notes page to the `mkdocs.yml` navigation and remove it from `not_in_nav`.
  - **Version Mentions**: Update all references to the new version throughout the documentation, including:
    - [Status in the root README](https://docs.orchardcore.net/en/latest/#status)
    - CLI templates and commands.
    - Relevant guides, such as the [Creating a new decoupled CMS Website](https://docs.orchardcore.net/en/latest/guides/decoupled-cms/) guide.
- [ ] Create a **version PR** titled `Release <version number>` (e.g., `Release 2.1.1`) from the version branch (e.g., `release/2.1.1`) into the release branch (e.g., `release/2.1`)
- [ ] In GitHub, manually run the `Preview - CI` workflow on your branch (NOT `main`). This will release a new preview version on Cloudsmith for testing.
```

## Step 3: Validation

- [ ] **Check Functionality**: Update [`OrchardCore.Samples`](https://github.com/OrchardCMS/OrchardCore.Samples) to the latest preview version generated in the previous step (just change the `OrchardCoreVersion` property in the root `Directory.Build.props` file). Ensure the samples work as expected.
- [ ] **Test Guides**: Test the following guides with NuGet packages from the Cloudsmith feed:
  - [Creating a modular ASP.NET Core application](https://docs.orchardcore.net/en/latest/guides/create-modular-application-mvc/)
  - [Creating an Orchard Core CMS website](https://docs.orchardcore.net/en/latest/guides/create-cms-application/)
  - [Creating a new decoupled CMS Website](https://docs.orchardcore.net/en/latest/guides/decoupled-cms/)
- [ ] If everything looks good and all checks pass, merge the version PR.

> ## Step 4: Creating the New Release

Reviewer: Create New Release is better for me. But whatever you like

Author: The template previously mixed the imperative mood of verb with nouns in titles (_create_ the new release/_align_ branches vs _backporting_ pull requests/code and documentation _updates_). Now it's all nouns. I'm fine with either of these, but I'd stick to one.

Reviewer: I am okay with sticking with one.

Author: That's the case now.

1. Navigate to the [GitHub Releases page](https://github.com/OrchardCMS/OrchardCore/releases/new).
2. In the "**Choose a tag**" menu, enter the new version number, including `v` (e.g., `v2.1.1`), and select "**+ Create tag: v... on publish**."
3. Change the target branch from `main` to the release branch (e.g., `release/2.1`).
4. Enter the version number in the Title field (e.g., `2.1.1`).
5. Click **Generate release notes** to auto-generate release notes.
6. Add a link to the release notes on the docs site: `Check out the full release notes [here](https://docs.orchardcore.net/en/latest/releases/2.1.1/).`
7. Ensure the "Set as the latest release" checkbox is checked, then click **Publish release**.

> ## Step 5: Aligning Branches

Reviewer: Same as above

> **Merge to `main`**: After releasing the new version, merge the release branch into the main branch to ensure `main` contains all administrative changes.
> 
> - [ ] Create a pull request from the release branch (e.g., `release/2.1`) into `main`.
> - [ ] Resolve any merge conflicts using external tools (e.g., Fork) to avoid auto-merging `main` into the release branch. **Important**: DO NOT resolve conflicts using GitHub's interface; that will automatically merge `main` into the release branch, which must be avoided.
> - [ ] Merge the PR if all checks pass. If there are merge conflicts, then you'll need to merge to `main` manually using the following steps:
>   1. Fetch the latest changes from the Git repository.
>   2. Checkout the `main` branch.
>   3. Merge the release branch (e.g., `release/2.1`) into `main` with a merge commit (NOT a squash merge). Use the commit message pattern `Merge release/2.1.1 to main`.
>   4. Resolve any conflicts.
>   5. Push the changes to `main`. This action requires a user with the ability to force-push into `main`, as it is protected by default.
>   6. GitHub will automatically delete the release branch; go back to the new merged PR to restore it.

Reviewer 2: ```suggestion
**Merge to `main`**: After releasing the new version, merge the release branch into the main branch to ensure that `main` contains all administrative changes.

- [ ] Create a pull request from the release branch (e.g., `release/2.1`) into `main`.
- [ ] Resolve any merge conflicts using external tools (e.g., Fork) to avoid auto-merging `main` into the release branch. **Important**: DO NOT resolve conflicts using GitHub's interface; that will automatically merge `main` into the release branch, which must be avoided.
- [ ] Merge the PR if all checks pass. If there are merge conflicts, then you'll need to merge to `main` manually using the following steps:
  1. Fetch the latest changes from the Git repository.
  2. Check out the `main` branch.
  3. Merge the release branch (e.g., `release/2.1`) into `main` with a merge commit (NOT a squash merge). Use the commit message pattern `Merge release/2.1.1 to main`.
  4. Resolve any conflicts.
  5. Push the changes to `main`. This action requires a user with the ability to force-push into `main`, as it is protected by default.
  6. GitHub will automatically delete the release branch; go back to the new merged PR to restore it.
```

Author: "Checkout" is a git technical term, not a misspelling. The other I added.

Reviewer 2: Right, but it depends in the context of running Git command, anyhow leave it

Author: How about now?

> - [ ] Create a pull request from the release branch (e.g., `release/2.1`) into `main`.
> - [ ] Resolve any merge conflicts using external tools (e.g., Fork) to avoid auto-merging `main` into the release branch. **Important**: DO NOT resolve conflicts using GitHub's interface; that will automatically merge `main` into the release branch, which must be avoided.
> - [ ] Merge the PR if all checks pass. If there are merge conflicts, then you'll need to merge to `main` manually using the following steps:
>   1. Fetch the latest changes from the Git repository.
>   2. Checkout the `main` branch.
>   3. Merge the release branch (e.g., `release/2.1`) into `main` with a merge commit (NOT a squash merge). Use the commit message pattern `Merge release/2.1.1 to main`.
>   4. Resolve any conflicts.
>   5. Push the changes to `main`. This action requires a user with the ability to force-push into `main`, as it is protected by default.
>   6. GitHub will automatically delete the release branch; go back to the new merged PR to restore it.

Reviewer 2: ```suggestion
- [ ] Create a pull request from the release branch (e.g., `release/2.1`) into `main`.
- [ ] Resolve any merge conflicts using external tools (e.g., Fork) to avoid auto-merging `main` into the release branch. **Important**: DO NOT resolve conflicts using GitHub's interface; that will automatically merge `main` into the release branch, which must be avoided.
- [ ] Merge the PR if all checks pass. If there are merge conflicts, then you'll need to merge to `main` manually using the following steps:
  1. Fetch the latest changes from the Git repository.
  2. `git checkout` the `main` branch.
  3. Merge the release branch (e.g., `release/2.1`) into `main` with a merge commit (NOT a squash merge). Use the commit message pattern `Merge release/2.1.1 to main`.
  4. Resolve any conflicts.
  5. Push the changes to `main`. This action requires a user with the ability to force-push into `main`, as it is protected by default.
  6. GitHub will automatically delete the release branch; to restore it, go back to the new merged PR.
```

## Step 6: Housekeeping

> - [ ] Assign the milestone for the release version to this issue. Create one if it doesn't exist.
> - [ ] Assign the milestone to the issues that were released in the version.
> - [ ] Close the milestone.
> - [ ] Update [`OrchardCore.Samples`](https://github.com/OrchardCMS/OrchardCore.Samples) to the new released version (just change the `OrchardCoreVersion` property in the root `Directory.Build.props` file).

Reviewer 2: ```suggestion
- [ ] Assign the milestone for the release version to this issue. Create one if it doesn't exist.
- [ ] Assign the milestone to the issues that were released in the version.
- [ ] Close the milestone.
- [ ] Update [`OrchardCore.Samples`](https://github.com/OrchardCMS/OrchardCore.Samples) to the newly released version (just change the `OrchardCoreVersion` property in the root `Directory.Build.props` file).
```

> ## Step 7: Publicizing the Release

Reviewer: Same
